### ISSUE TITLE
ganeti-mond: ERROR 500: Internal Server Error.

### ISSUE THREAD

**Issue by Ganeti-Issues-Migrator** [ganeti/ganeti#925]
Originally reported of Google Code with ID 869.

```
What software version are you running? Please provide the output of "gnt-
cluster --version", "gnt-cluster version", and "hspace --version".

<b>What distribution are you using?</b>
root@node2 ~ # gnt-cluster --version
gnt-cluster (ganeti v2.10.5) 2.10.5
root@node2 ~ # gnt-cluster version
Software version: 2.10.5
Internode protocol: 2100000
Configuration format: 2100000
OS api version: 20
Export interface: 0
VCS version: (ganeti) version v2.10.5
root@node2 ~ # hspace --version
hspace (ganeti) version v2.10.5
compiled with ghc 7.4
running on linux x86_64
root@node2 ~ # cat /etc/debian_version 
7.5
root@node2 ~ # dpkg -l ganeti
ganeti    2.10.5-1~bpo70+1


<b>What steps will reproduce the problem?</b>
a. wget -O - http://127.0.0.1:1815/1/report/all
or
b. wget -O - http://127.0.0.1:1815/1/report/storage/drbd

<b>What is the expected output? What do you see instead?</b>
<root@node2 ~log/ganeti # wget -O - http://127.0.0.1:1815/1/report/storage/drbd                  
<--2014-07-02 19:15:55--  http://127.0.0.1:1815/1/report/storage/drbd
<Connecting to 127.0.0.1:1815... connected.
<HTTP request sent, awaiting response... 500 Internal Server Error
<2014-07-02 19:15:55 ERROR 500: Internal Server Error.
> data

<b>Please provide any additional information below.</b>
root@node2 ~log/ganeti # tail -f monitoring-daemon-error.log 
[02/Jul/2014:19:14:01 +0200] During processing of request from 127.0.0.1:52138
request:
"Request <\n    server-name: 127.0.0.1\n    remote: 127.0.0.1:52138\n    local: 127.0.0.1:1815\n    Headers:\n      ========================================\n          User-Agent: Wget/1.13.4 (linux-gnu)\n Connection: Keep-Alive\n Host: 127.0.0.1:1815\n Accept: */*\n      ========================================\n    content-length: Nothing\n    method: GET\n    version: (1,1)\n    cookies:\n      ========================================\n      \n      ========================================\n    pathinfo: 1/report/all\n    contextpath: /\n    URI: /1/report/all\n    params:\n      ========================================\n      \n      ========================================\n>"
A web handler threw an exception. Details:
ExitFailure 1
[02/Jul/2014:19:14:26 +0200] During processing of request from 127.0.0.1:52139
request:
"Request <\n    server-name: 127.0.0.1\n    remote: 127.0.0.1:52139\n    local: 127.0.0.1:1815\n    Headers:\n      ========================================\n          User-Agent: Wget/1.13.4 (linux-gnu)\n Connection: Keep-Alive\n Host: 127.0.0.1:1815\n Accept: */*\n      ========================================\n    content-length: Nothing\n    method: GET\n    version: (1,1)\n    cookies:\n      ========================================\n      \n      ========================================\n    pathinfo: 1/report/storage/drbd\n    contextpath: /\n    URI: /1/report/storage/drbd\n    params:\n      ========================================\n      \n      ========================================\n>"
A web handler threw an exception. Details:
ExitFailure 1

```
Originally added on 2014-07-02 17:18:00 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
I just tried and failed to reproduce it on a clean stable-2.10 checkout.

So we would need more information on the specifics. I assign the issue to
Michele who did most of the work on the monitoring daemon and CC Apollon
who is the maintainer for the debian package.

```
Originally added on 2014-07-03 13:07:31 +0000 UTC.

Changed State: NeedFeedback

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi Neal. Your usage of the daemon seems indeed to be correct, and this output is not providing much help.

Could you please have a look at the monitoring daemon logs, to see if they contain any useful information?

Usually they are in /var/log/ganeti/monitoring-daemon*.log

Cheers,
Michele
```
Originally added on 2014-07-03 14:01:49 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
-- Empty comment --
```
Originally added on 2014-07-03 14:03:27 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Hi Michele,

Running mond in the foreground with debugging on, I get:

Error: "\n 1: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----\n    ns:0 nr:0 dw:0 dr:744 al:0 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0\n\n 3: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----\n    ns:0 nr:0 dw:0 dr:664 al:0 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0\n"
[]
Failed reading: takeWhile1

So it seems something's wrong with the DRBD collector.

Cheers,
Apollon
```
Originally added on 2014-07-03 14:05:19 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
there isn't more than that what I added as additional information (only the same error(s) with other timestamps, and the access.log)

127.0.0.1 - - [02/Jul/2014:19:15:23 +0200] "GET /1/report/instance/inst-status-xen HTTP/1.1" 200 - - "Wget/1.13.4 (linux-gnu)"
127.0.0.1 - - [02/Jul/2014:19:15:55 +0200] "GET /1/report/storage/drbd HTTP/1.1" 500 56 - "Wget/1.13.4 (linux-gnu)"
```
Originally added on 2014-07-03 14:07:57 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Thanks Apollon, thats quite interesting! So, it's probably an actual bug. /proc/drbd's format continuously changes, and maybe there is something new we have to support. For example, without looking at the source code, that "\n" at the beginning of the line does not sound familiar to me...

Would you mind attaching here (or sending me) a copy of a /proc/drbd file that makes the data collector crash?
```
Originally added on 2014-07-04 07:43:30 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
version: 8.4.3 (api:1/proto:86-101)
srcversion: F97798065516C94BE0F27DC

 1: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:0 nr:0 dw:0 dr:744 al:0 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0

 3: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:0 nr:0 dw:0 dr:664 al:0 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0


Removing the empty line before 1: makes the parser happy. Presumably it doesn't like the resources stanza to start with a blank line (which is perfectly normal if minor 0 has been freed).

Note that empty lines within the resource stanza are consumed by

          additionalEOL = A.skipWhile A.isEndOfLine

and that's probably why the empty line between 1: and 3: does not break the parser.
```
Originally added on 2014-07-04 08:11:10 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
version: 8.3.11 (api:88/proto:86-96)
srcversion: F937DCB2E5D83C6CCE4A6C9 

 1: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:2735476 nr:0 dw:2735476 dr:22461460 al:82 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
 2: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:3893084 nr:0 dw:3893084 dr:15748312 al:489 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
 3: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:2536404 nr:0 dw:2536832 dr:23985536 al:53 bm:6 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0

 5: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:22799456 nr:0 dw:56354628 dr:30352340 al:4288 bm:2071 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
 6: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:181870664 nr:0 dw:181928504 dr:188828668 al:95729 bm:37 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
 7: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:1571985684 dw:1571985684 dr:0 al:0 bm:1023 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
 8: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:225822732 nr:0 dw:225822732 dr:106676644 al:38086 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
 9: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:51543192 dw:51543192 dr:0 al:0 bm:2047 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
10: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:318117896 nr:0 dw:318169208 dr:45630832 al:21140 bm:31 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
11: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:376236 nr:0 dw:376292 dr:24332588 al:20 bm:1 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
12: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:215650600 dw:215650600 dr:0 al:0 bm:4082 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
13: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:35485172 dw:35485172 dr:0 al:0 bm:2048 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
14: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:49664040 nr:0 dw:49664152 dr:29264892 al:2730 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
15: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:6314448 nr:0 dw:6314448 dr:29495756 al:1590 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
16: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:80105968 dw:80105968 dr:0 al:0 bm:512 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
17: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:358615228 dw:358615228 dr:0 al:0 bm:2046 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
18: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:186907932 dw:186907932 dr:0 al:0 bm:2048 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
19: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:0 nr:0 dw:0 dr:23989352 al:0 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
20: cs:Connected ro:Secondary/Primary ds:UpToDate/UpToDate C r-----
    ns:0 nr:221385104 dw:221385104 dr:0 al:0 bm:4096 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
21: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:6366900 nr:0 dw:6367248 dr:24172184 al:335 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
22: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:33239672 nr:0 dw:33239816 dr:23983960 al:1034 bm:5 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
23: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:134217728 nr:0 dw:0 dr:157223292 al:0 bm:8190 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0
24: cs:Connected ro:Primary/Secondary ds:UpToDate/UpToDate C r-----
    ns:1295412 nr:0 dw:1295412 dr:19152460 al:77 bm:0 lo:0 pe:0 ua:0 ap:0 ep:1 wo:f oos:0


here mine, it looks a bit different (but only with the line-setup and actual data from a cluster)
```
Originally added on 2014-07-04 08:29:02 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Neal is also lacking the first DRBD resource, which confirms my suspicion. I'm preparing a patch set, will send it to the list.

Apollon
```
Originally added on 2014-07-04 08:49:47 +0000 UTC.

**Comment by Ganeti-Issues-Migrator** [comment]
```
Ok, thanks a lot!
```
Originally added on 2014-07-04 09:06:03 +0000 UTC.

Changed State: Accepted

**Comment by Ganeti-Issues-Migrator** [comment]
```
commit 40d79e735b0341c8322240cb2548ee021eaacf85
Author: Apollon Oikonomopoulos <apoikos@gmail.com>
Date:   Fri Jul 4 12:20:17 2014 +0300

    DRBD parser: consume initial empty resource lines
    
    When a DRBD resource is taken down, its minor number is freed and its
    place in /proc/drbd is taken by an empty line. Currently the parser is
    able to consume any empty lines appearing inside the resource stanza,
    but cannot handle the first resource being empty.
    
    We fix this by consuming all EOL characters appearing before the device
    number. A test for this case should also be added.
    
    This fixes issue #869.
    
    Signed-off-by: Apollon Oikonomopoulos <apoikos@gmail.com>
    Signed-off-by: Klaus Aehlig <aehlig@google.com>
    Reviewed-by: Klaus Aehlig <aehlig@google.com>

```
Originally added on 2014-07-04 09:32:19 +0000 UTC.

Changed State: Fixed
Added to Milestone: Release2.8

**Comment by Ganeti-Issues-Migrator** [comment]
```
Released as part of 2.14.0 (and probably a lot earlier releases)
```
Originally added on 2015-06-02 10:42:35 +0000 UTC.

Changed State: Released
